### PR TITLE
Initial Stargate support (wip)

### LINKED ISSUE #5: Create initial CRD and controller for Stargate
For this ticket we want to introduce a minimal CRD and initial controller. Assume a Stargate instance per CassandraDatacenter by default for now.

Integration and e2e tests should be added/updated as necessary.

We can worry about filling out the CRD in follow up tickets.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-581) by [Unito](https://www.unito.io)
┆Epic: Stargate Controller Integration
┆Fix Versions: k8ssandra-2.0.0-alpha

### PR DESCRIPTION
**Pull request by adutra**
**What this PR does**:
Adds initial support for Stargate CRD and controller.

**Which issue(s) this PR fixes**:
Fixes #5 

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: api/v1alpha1/stargate_types.go
Comment (jsanda) [inline api/v1alpha1/stargate_types.go:59]: Keep in mind that a pod can have multiple init containers. If this is for the init container we use in the Helm, then I would include that in the field name. With that said, we won't need that init container :) The purpose of the init container we deploy with the chart is to wait to deploy Stargate until the Cassandra cluster is up. The K8ssandraCluster controller can do this. It will be the one that creates the Stargate and Reaper objects. It should wait to deploy them until the Datacenter is up. Of course, both Stargate and Reaper can be deployed without a K8ssandraCluster, so I would include logic in the Stargate controller to check that the datacenter is ready before deploying Stargate.

File: api/v1alpha1/stargate_types.go
Comment (jsanda) [inline api/v1alpha1/stargate_types.go:88]: Based on prior discussions with @adejanovski my understanding is that we want to deploy Stargate nodes in each rack. Affinity for Cassandra pods is configured implicitly via racks. I think we will want piggy back off of that. By default, I think we should look at the rack configuration of the CassandraDatacenter and configure affinity such that we will get Stargate pods distributed evenly among racks (as best we can). This goes beyond the scope of this PR, but I wanted to at least point it out here.

**Review by jsanda — COMMENTED** [review]
I haven't made it through all the code yet but it looks good!

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:68]: Before doing anything else you should do `stargate = stargate.DeepCopy()`. Reads, i.e., `Get()` hit a shared cached. We don't want to modify objects in the cache.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:80]: It could also be that the CassandraDatacenter has not been created.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:151]: As I mentioned in an earlier comment, we won't need this init container. The Stargate controller can wait to create the Stargate Deployment until the CassandraDatacenter is ready.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:88]: A common pattern is to create the object, compute the hash, and store it in an annotation. Then when the object already exists, you can simply compare hashes to see if there are changes. You can find an example of this in k8ssandracluster_controller.go. You will also find this in cass-operator with the StatefulSets that it creates.

**Review by jsanda — CHANGES_REQUESTED** [review]

File: api/v1alpha1/stargate_types.go
Comment (adutra) [reply api/v1alpha1/stargate_types.go:49]: You are right that this field is only relevant if `ContainerImage` is nil, which does feel a bit like a code smell. I agree to remove it.

File: controllers/stargate_controller.go
Comment (adutra) [reply controllers/stargate_controller.go:68]: Ok I will. That said, the controller doesn't change the `Stargate` object, it only updates the `Deployment` object. But that might change in the future I guess, so better safe than sorry.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:90]: We also need to create the Service object for Stargate.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:216]: I don't think we need the DeepCopy `stargate` is already a copy.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:227]: I don't think we need the DeepCopy `stargate` is already a copy.

File: controllers/stargate_controller.go
Comment (jsanda) [inline controllers/stargate_controller.go:150]: I would make `newStargateDeployment` a stand alone function instead of a method of StargateReconciler. This will make it a lot easier for writing unit tests.

File: test/framework/e2e_framework.go
Comment (jsanda) [inline test/framework/e2e_framework.go:278]: We are starting to see a pattern across multiple functions that involve performing some action across cluster. Then poll across all clusters until condition is met. Maybe we can extract some of the control flow parts into separate functions. Doesn't have to be done here in this PR. Just thinking out loud...

File: test/e2e/suite_test.go
Comment (jsanda) [inline test/e2e/suite_test.go:25]: After thinking about it some more I do think it would be worthwhile to have some e2e test coverage for deployment Stargate on its own without a parent K8ssandraCluster. We still need a CassandraDatacenter though. It allows us to get the full test coverage of the Stargate controller without the complexity and overhead of the multi-cluster stuff. These tests would only require a single k8s cluster and could be run with the free tier GH Actions runners, which is a big plus.

wdyt?

**Review by jsanda — CHANGES_REQUESTED** [review]

File: test/e2e/suite_test.go
Comment (jsanda) [reply test/e2e/suite_test.go:25]: Of course, we should have no trouble running a multi-cluster env with only a single cluster with the free tier runner in GH Actions.

**Review by jsanda — CHANGES_REQUESTED** [review]
I rebased your branch on top of my branch for #50 to do some testing. My work is here https://github.com/jsanda/k8ssandra-operator/tree/stargate-test. I am seeing this error after the first Stargate instance is deployed:

`2021-07-26T02:04:26.344Z	ERROR	controller-runtime.manager.controller.k8ssandracluster	Reconciler error	{"reconciler group": "k8ssandra.io", "reconciler kind": "K8ssandraCluster", "name": "test", "namespace": "multi-dc-tns82l", "error": "stargates.k8ssandra.io \"test-dc1-stargate\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update"}`

It also looks like the resource annotation on the Stargate object is lost along the way.

File: controllers/k8ssandracluster_controller.go
Comment (adutra) [reply controllers/k8ssandracluster_controller.go:147]: Ah right, I forgot that possibility. OK to leave it up to another PR.

**Comment by adutra** [comment]
> I am seeing this error after the first Stargate instance is deployed [...] metadata.resourceVersion: Invalid value: 0x0: must be specified for an update

I have not seen this error in my own branch, but I admit that the error pops up in yours. My understanding is that these two lines are maybe doing more harm than good:

```
actualStargate = actualStargate.DeepCopy()
desiredStargate.DeepCopyInto(actualStargate)
...
remoteClient.Update(ctx, actualStargate)					
```

I changed this to:

```
desiredStargate.SetResourceVersion(actualStargate.GetResourceVersion())
...
remoteClient.Update(ctx, desiredStargate)					
```

And the error disappears. Do you know what's the preferred idiom to update an existing `actual` resource so that it reflects the `desired` one? 

> It also looks like the resource annotation on the Stargate object is lost along the way.

I haven't been able to reproduce the situation where the resource hash annotation is lost during updates.

**Comment by jsanda** [comment]
>And the error disappears. Do you know what's the preferred idiom to update an existing actual resource so that it reflects the desired one?

Great question. See [here](https://github.com/k8ssandra/cass-operator/blob/master/pkg/reconciliation/reconcile_racks.go#L244) in cass-operator where it uses the `DeepCopyInto` function. It also calls `SetResourceVersion`.

**Review by jsanda — APPROVED** [review]
@adutra please feel free to merge when you are ready. And of course I am happy review/test more before merging if you want. Considering we are still in a prototyping phase and that I have done testing with you branch, I think we are good to merge to address changes in follow up PRs.

**Comment by adutra** [comment]
@jsanda I just finished rebasing and re-testing everything, so I'm OK too to merge now. Among the topics we discussed for follow-up PRs:

# Ability to delete Stargate resources from k8ssandracluster controller;
# Revisit / fix Stargate test on multi-cluster scenario.

Should I go ahead and create tickets for the above topics?

**Pull request merged**
